### PR TITLE
[GSK-1955] stdout of scan

### PR DESCRIPTION
**Pull request by rabah-khalek**

### CODE REVIEW

**Comment by mattbit** [comment]
I’m getting this error @rabah-khalek when the scan has no issues:

```
File [~/Code/giskard/giskard/scanner/__init__.py:61](https://file+.vscode-resource.vscode-cdn.net/Users/matteo/Code/llm-scan-testing/~/Code/giskard/giskard/scanner/__init__.py:61), in scan(model, dataset, params, only, verbose, raise_exceptions)
ref='~/Code/giskard/giskard/scanner/__init__.py:1'>1</a>;32m     [34](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=33) """Automatically detects model vulnerabilities.
ref='~/Code/giskard/giskard/scanner/__init__.py:1'>1</a>;32m     [35](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=34) 
ref='~/Code/giskard/giskard/scanner/__init__.py:1'>1</a>;32m     [36](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=35) See :class:`Scanner` for more details.
ref='~/Code/giskard/giskard/scanner/__init__.py:0'>0</a>;32m   (...)
ref='~/Code/giskard/giskard/scanner/__init__.py:1'>1</a>;32m     [58](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=57)     A scan report object containing the results of the scan.
ref='~/Code/giskard/giskard/scanner/__init__.py:1'>1</a>;32m     [59](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=58) """
ref='~/Code/giskard/giskard/scanner/__init__.py:1'>1</a>;32m     [60](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=59) scanner = Scanner(params, only=only)
ref='~/Code/giskard/giskard/scanner/__init__.py:0'>0</a>;32m---> [61](file:///Users/matteo/Code/giskard/giskard/scanner/__init__.py?line=60) return scanner.analyze(model, dataset, verbose=verbose, raise_exceptions=raise_exceptions)

File [~/Code/giskard/giskard/scanner/scanner.py:135](https://file+.vscode-resource.vscode-cdn.net/Users/matteo/Code/llm-scan-testing/~/Code/giskard/giskard/scanner/scanner.py:135), in Scanner.analyze(self, model, dataset, verbose, raise_exceptions)
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [132](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=131) elapsed = perf_counter() - time_start
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [134](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=133) if verbose:
ref='~/Code/giskard/giskard/scanner/scanner.py:0'>0</a>;32m--> [135](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=134)     self._print_execution_summary(model, issues, errors, elapsed)
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [137](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=136) self._collect_analytics(model, dataset, issues, elapsed, model_validation_time)
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [139](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=138) return ScanReport(issues, model=model, dataset=dataset)

File [~/Code/giskard/giskard/scanner/scanner.py:356](https://file+.vscode-resource.vscode-cdn.net/Users/matteo/Code/llm-scan-testing/~/Code/giskard/giskard/scanner/scanner.py:356), in Scanner._print_execution_summary(self, model, issues, errors, elapsed)
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [354](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=353) if model.is_text_generation:
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [355](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=354)     measured = self._get_cost_measure()
ref='~/Code/giskard/giskard/scanner/scanner.py:0'>0</a>;32m--> [356](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=355)     xprint("Cost summary…", *measured.values(), template=Catalog.ScanCostSummary)
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [357](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=356) if errors:
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [358](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=357)     warning(
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [359](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=358)         f"{len(errors)} errors were encountered while running detectors. Please check the log to understand what went wrong. "
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [360](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=359)         "You can run the scan again with `raise_exceptions=True` to disable graceful handling."
ref='~/Code/giskard/giskard/scanner/scanner.py:1'>1</a>;32m    [361](file:///Users/matteo/Code/giskard/giskard/scanner/scanner.py?line=360)     )

File [~/Code/giskard/giskard/utils/xprint.py:156](https://file+.vscode-resource.vscode-cdn.net/Users/matteo/Code/llm-scan-testing/~/Code/giskard/giskard/utils/xprint.py:156), in xprint(template, filename, verbose, *args, **kwargs)
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m    [153](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=152) if not verbose:
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m    [154](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=153)     return
ref='~/Code/giskard/giskard/utils/xprint.py:0'>0</a>;32m--> [156](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=155) args = style(*args, template=template)
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m    [158](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=157) if filename is not None:
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m    [159](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=158)     with open(filename, "a") as f:

File [~/Code/giskard/giskard/utils/xprint.py:77](https://file+.vscode-resource.vscode-cdn.net/Users/matteo/Code/llm-scan-testing/~/Code/giskard/giskard/utils/xprint.py:77), in style(template, *args)
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m     [75](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=74)         out += [wraps[i], *args[i:]]
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m     [76](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=75)         break
ref='~/Code/giskard/giskard/utils/xprint.py:0'>0</a>;32m---> [77](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=76)     out += [wraps[i], args[i]]
     [78](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=77) out += [wraps[-[1](https://file+.vscode-resource.vscode-cdn.net/Users/matteo/Code/llm-scan-testing/~/Code/giskard/giskard/utils/xprint.py:1)]]
ref='~/Code/giskard/giskard/utils/xprint.py:1'>1</a>;32m     [79](file:///Users/matteo/Code/giskard/giskard/utils/xprint.py?line=78) return out

IndexError: tuple index out of range
```

**Comment by rabah-khalek** [comment]
Thanks @mattbit I'll have a look

File: giskard/utils/xprint.py
Comment (mattbit) [inline giskard/utils/xprint.py:1]: Let’s find a better module for this. `utils` too generic. We already have a `visualization` module, maybe this should also go there?

File: giskard/utils/xprint.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+Estimated calls to your model: ~{}
+Estimated OpenAI GPT-4 calls for evaluation: {} (~{} prompt tokens and ~{} sampled tokens)
+OpenAI API costs for evaluation are estimated to ${}.
+"""
+
+SCAN_COST_SUMMARY_CONTENT = """💵 {}
+LLM-assisted detectors have used the following resources:
+OpenAI GPT-4 calls for evaluation: {} ({} prompt tokens and {} sampled tokens)
+OpenAI API costs for evaluation amount to ${} (standard pricing).
+"""
```
Comment (mattbit) [inline giskard/utils/xprint.py:104]: Why is this stuff in `utils`? This should be in the `scanner` module

File: giskard/utils/xprint.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        content="-" * (CHARS_LIMIT // 2 - 8) + " Summary of {} " + "-" * (CHARS_LIMIT // 2 - 8), pstyles=[BLUE_STYLE]
+    )
+    DetectedIssues = Template(content="{}: {} detected. (Took {})", pstyles=[BLUE_STYLE, RED_STYLE, MAGENTA_STYLE])
+    NoDetectedIssues = Template(content="{}: {} detected. (Took {})", pstyles=[BLUE_STYLE, GREEN_STYLE, MAGENTA_STYLE])
+    ScanCostEstimate = Template(
+        content=SCAN_COST_ESTIMATE_CONTENT,
+        pstyles=[YELLOW_STYLE, MAGENTA_STYLE, MAGENTA_STYLE, CYAN_STYLE, CYAN_STYLE, RED_STYLE],
+    )
+    ScanCostSummary = Template(
+        content=SCAN_COST_SUMMARY_CONTENT,
```
Comment (mattbit) [inline giskard/utils/xprint.py:141]: Same. I would put the templates where they belong. Part of this belongs to the `Scanner`, other to the `Detector` instances, etc. I don’t think that the grouping on the `Catalog` is meaningful or neecessary.

File: giskard/utils/xprint.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    filename: Optional[str] = None,
+    verbose: bool = True,
+    **kwargs,
+):
+    if not verbose:
+        return
+
+    args = style(*args, template=template)
+
+    if filename is not None:
```
Comment (mattbit) [inline giskard/utils/xprint.py:158]: We may want to strip coloring if printing to file

**Review by mattbit — CHANGES_REQUESTED** [review]
First pass:
- Let’s try to keep interdependencies with other modules minimal (no output code in `llm` module, no `verbose` passed downstream from scan to Scanner to Detector to Evaluator)
- Having the centralized `Catalog` dataclass is not useful and increases interdependencies
- Can we make the output coloring less "Christmas tree", using colors more sparingly and more consistent? (Have a look at how `poetry`, `pdm` or others use colors, sometimes with a lot of colors but very consistently, e.g. [here](https://raw.githubusercontent.com/bloomberg/memray/main/docs/_static/images/live_animated.gif))
- There are still several errors
- I couldn’t run it fully, but can we find a very standardized way to do the detectors logging with clear PASS/FAIL results or similar? Garak does it nicely. It would be a major improvement with respect to the current logging we have if we could really separate what is "debugging" information and what is the scan result as it gets collected.

---

On the implementation, you could replace the `logger` (in `scanner.logger.logger`) with a custom class that handles this logging, maybe with special methods like `log_detector_result` or similar. I believe this could help in making it more modular.

**Comment by mattbit** [comment]
FYI other error I’m getting: `TypeError: RequirementBasedDetector.run() got an unexpected keyword argument 'verbose'`

**Pull request closed without merging**